server/http.go

```diff
@@ -66,27 +71,69 @@ func handleHTTPRequest(p *Plugin, w http.ResponseWriter, r *http.Request) (int,
 		return httpACUninstalled(p, w, r)
 
 	// Atlassian Connect user mapping
-	case routeACUserConfig:
-		return httpACUserConfig(p, w, r)
-	case routeACUserConfigSubmit:
-		return httpACUserConfigSubmit(p, w, r)
+	case routeACUserRedirectWithToken:
+		return withCloudInstance(p, w, r, httpACUserRedirect)
+	case routeACUserConfirm,
+		routeACUserConnected,
+		routeACUserDisconnected:
+		return withCloudInstance(p, w, r, httpACUserInteractive)
 
 	// Incoming webhook
 	case routeIncomingWebhook, routeIncomingIssueEvent:
 		return httpWebhook(p, w, r)
 
 	// Oauth1 (Jira Server)
 	case routeOAuth1Complete:
-		return httpOAuth1Complete(p, w, r)
+		return withServerInstance(p, w, r, httpOAuth1Complete)
 	case routeOAuth1PublicKey:
 		return httpOAuth1PublicKey(p, w, r)
 
 	// User connect/disconnect links
 	case routeUserConnect:
-		return httpUserConnect(p, w, r)
+		return withInstance(p, w, r, httpUserConnect)
 	case routeUserDisconnect:
-		return httpUserDisconnect(p, w, r)
+		return withInstance(p, w, r, httpUserDisconnect)
 	}
 
 	return http.StatusNotFound, errors.New("not found")
 }
+
+func (p *Plugin) loadTemplates(dir string) (map[string]*template.Template, error) {
+	templates := make(map[string]*template.Template)
+	err := filepath.Walk(dir, func(path string, info os.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+		if info.IsDir() {
+			return nil
+		}
+		template, err := template.ParseFiles(path)
+		if err != nil {
+			p.errorf("OnActivate: failed to parse template %s: %v", path, err)
+			return nil
+		}
+		key := path[len(dir):]
+		templates[key] = template
+		p.debugf("loaded template %s", key)
+		return nil
+	})
+	if err != nil {
+		return nil, errors.WithMessage(err, "OnActivate: failed to load templates")
+	}
+	return templates, nil
+}
+
+func (p *Plugin) respondWithTemplate(w http.ResponseWriter, r *http.Request, contentType string, values interface{}) (int, error) {
+	w.Header().Set("Content-Type", contentType)
```

Comment:
We can probably detect this at load time rather than having to pass a parameter.